From 6347cd3e72347d8698af0b48fc626fa43f4e218f Mon Sep 17 00:00:00 2001
From: Arun Raghavan <arun@arunraghavan.net>
Date: Mon, 4 Jul 2016 13:57:16 +0530
Subject: [PATCH 09/18] pulsecore: Move pa_core structure into its own header

The idea is to allow some parts of the code to use pa_core as an opaque
structure and access required members via API, over which we can then
perform some form of access control

Signed-off-by: Arun Raghavan <arun@arunraghavan.net>
---
 src/Makefile.am                                 |   1 +
 src/daemon/main.c                               |   1 +
 src/daemon/server-lookup.c                      |   1 +
 src/modules/alsa/alsa-sink.c                    |   1 +
 src/modules/alsa/alsa-source.c                  |   1 +
 src/modules/alsa/alsa-ucm.c                     |   1 +
 src/modules/alsa/module-alsa-card.c             |   1 +
 src/modules/bluetooth/backend-native.c          |   1 +
 src/modules/bluetooth/bluez4-util.c             |   1 +
 src/modules/bluetooth/bluez5-util.c             |   1 +
 src/modules/bluetooth/module-bluetooth-policy.c |   1 +
 src/modules/bluetooth/module-bluez4-device.c    |   1 +
 src/modules/bluetooth/module-bluez5-device.c    |   1 +
 src/modules/bluetooth/module-bluez5-discover.c  |   1 +
 src/modules/dbus/iface-card.c                   |   1 +
 src/modules/dbus/iface-client.c                 |   1 +
 src/modules/dbus/iface-core.c                   |   1 +
 src/modules/dbus/iface-device-port.c            |   1 +
 src/modules/dbus/iface-device.c                 |   1 +
 src/modules/dbus/iface-memstats.c               |   1 +
 src/modules/dbus/iface-module.c                 |   1 +
 src/modules/dbus/iface-sample.c                 |   1 +
 src/modules/dbus/iface-stream.c                 |   1 +
 src/modules/dbus/module-dbus-protocol.c         |   1 +
 src/modules/echo-cancel/adrian.c                |   1 +
 src/modules/echo-cancel/module-echo-cancel.c    |   1 +
 src/modules/gconf/module-gconf.c                |   1 +
 src/modules/jack/module-jack-sink.c             |   1 +
 src/modules/jack/module-jack-source.c           |   1 +
 src/modules/module-access.c                     |   1 +
 src/modules/module-allow-passthrough.c          |   1 +
 src/modules/module-always-sink.c                |   1 +
 src/modules/module-augment-properties.c         |   1 +
 src/modules/module-card-restore.c               |   1 +
 src/modules/module-cli.c                        |   1 +
 src/modules/module-combine-sink.c               |   1 +
 src/modules/module-default-device-restore.c     |   1 +
 src/modules/module-device-manager.c             |   1 +
 src/modules/module-device-restore.c             |   1 +
 src/modules/module-equalizer-sink.c             |   1 +
 src/modules/module-esound-sink.c                |   1 +
 src/modules/module-filter-apply.c               |   1 +
 src/modules/module-filter-heuristics.c          |   1 +
 src/modules/module-intended-roles.c             |   1 +
 src/modules/module-ladspa-sink.c                |   1 +
 src/modules/module-lirc.c                       |   1 +
 src/modules/module-loopback.c                   |   1 +
 src/modules/module-match.c                      |   1 +
 src/modules/module-mmkbd-evdev.c                |   1 +
 src/modules/module-native-protocol-fd.c         |   1 +
 src/modules/module-null-sink.c                  |   1 +
 src/modules/module-null-source.c                |   1 +
 src/modules/module-pipe-sink.c                  |   1 +
 src/modules/module-pipe-source.c                |   1 +
 src/modules/module-position-event-sounds.c      |   1 +
 src/modules/module-protocol-stub.c              |   1 +
 src/modules/module-rescue-streams.c             |   1 +
 src/modules/module-role-cork.c                  |   1 +
 src/modules/module-role-ducking.c               |   1 +
 src/modules/module-rygel-media-server.c         |   1 +
 src/modules/module-sine-source.c                |   1 +
 src/modules/module-sine.c                       |   1 +
 src/modules/module-stream-restore.c             |   1 +
 src/modules/module-suspend-on-idle.c            |   1 +
 src/modules/module-switch-on-connect.c          |   1 +
 src/modules/module-switch-on-port-available.c   |   1 +
 src/modules/module-systemd-login.c              |   1 +
 src/modules/module-tunnel-sink-new.c            |   1 +
 src/modules/module-tunnel-source-new.c          |   1 +
 src/modules/module-tunnel.c                     |   1 +
 src/modules/module-udev-detect.c                |   1 +
 src/modules/module-virtual-sink.c               |   1 +
 src/modules/module-virtual-source.c             |   1 +
 src/modules/module-virtual-surround-sink.c      |   1 +
 src/modules/module-zeroconf-discover.c          |   1 +
 src/modules/module-zeroconf-publish.c           |   1 +
 src/modules/oss/module-oss.c                    |   1 +
 src/modules/raop/module-raop-discover.c         |   1 +
 src/modules/raop/module-raop-sink.c             |   1 +
 src/modules/raop/raop_client.c                  |   1 +
 src/modules/rtp/module-rtp-recv.c               |   1 +
 src/modules/rtp/module-rtp-send.c               |   1 +
 src/modules/stream-interaction.c                |   1 +
 src/modules/x11/module-x11-cork-request.c       |   1 +
 src/modules/x11/module-x11-xsmp.c               |   1 +
 src/pulsecore/card.c                            |   1 +
 src/pulsecore/cli-command.c                     |   1 +
 src/pulsecore/cli-text.c                        |   1 +
 src/pulsecore/client.c                          |   1 +
 src/pulsecore/core-scache.c                     |   1 +
 src/pulsecore/core-struct.h                     | 103 ++++++++++++++++++++++++
 src/pulsecore/core-subscribe.c                  |   1 +
 src/pulsecore/core.c                            |   1 +
 src/pulsecore/core.h                            |  76 -----------------
 src/pulsecore/dbus-shared.c                     |   1 +
 src/pulsecore/device-port.c                     |   1 +
 src/pulsecore/module.c                          |   1 +
 src/pulsecore/namereg.c                         |   1 +
 src/pulsecore/play-memchunk.c                   |   1 +
 src/pulsecore/protocol-esound.c                 |   1 +
 src/pulsecore/protocol-http.c                   |   1 +
 src/pulsecore/protocol-native.c                 |   1 +
 src/pulsecore/protocol-simple.c                 |   1 +
 src/pulsecore/shared.c                          |   1 +
 src/pulsecore/sink-input.c                      |   1 +
 src/pulsecore/sink.c                            |   1 +
 src/pulsecore/sound-file-stream.c               |   1 +
 src/pulsecore/source-output.c                   |   1 +
 src/pulsecore/source.c                          |   1 +
 src/pulsecore/x11wrap.c                         |   1 +
 110 files changed, 211 insertions(+), 76 deletions(-)
 create mode 100644 src/pulsecore/core-struct.h

diff --git a/src/Makefile.am b/src/Makefile.am
index 7c66064..70bc848 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -960,6 +960,7 @@ libpulsecore_@PA_MAJORMINOR@_la_SOURCES = \
 		pulsecore/core-scache.c pulsecore/core-scache.h \
 		pulsecore/core-subscribe.c pulsecore/core-subscribe.h \
 		pulsecore/core.c pulsecore/core.h \
+		pulsecore/core-struct.h \
 		pulsecore/hook-list.c pulsecore/hook-list.h \
 		pulsecore/ltdl-helper.c pulsecore/ltdl-helper.h \
 		pulsecore/modargs.c pulsecore/modargs.h \
diff --git a/src/daemon/main.c b/src/daemon/main.c
index dc5e5bc..cb69f17 100644
--- a/src/daemon/main.c
+++ b/src/daemon/main.c
@@ -73,6 +73,7 @@
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/module.h>
 #include <pulsecore/cli-command.h>
 #include <pulsecore/log.h>
diff --git a/src/daemon/server-lookup.c b/src/daemon/server-lookup.c
index ca390c7..b5c53e4 100644
--- a/src/daemon/server-lookup.c
+++ b/src/daemon/server-lookup.c
@@ -27,6 +27,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-shared.h>
 #include <pulsecore/macro.h>
diff --git a/src/modules/alsa/alsa-sink.c b/src/modules/alsa/alsa-sink.c
index 886c735..41ef1ae 100644
--- a/src/modules/alsa/alsa-sink.c
+++ b/src/modules/alsa/alsa-sink.c
@@ -38,6 +38,7 @@
 #include <pulse/internal.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/module.h>
 #include <pulsecore/memchunk.h>
diff --git a/src/modules/alsa/alsa-source.c b/src/modules/alsa/alsa-source.c
index b788df2..4a3f407 100644
--- a/src/modules/alsa/alsa-source.c
+++ b/src/modules/alsa/alsa-source.c
@@ -33,6 +33,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/module.h>
 #include <pulsecore/memchunk.h>
diff --git a/src/modules/alsa/alsa-ucm.c b/src/modules/alsa/alsa-ucm.c
index b42c040..227ed7d 100644
--- a/src/modules/alsa/alsa-ucm.c
+++ b/src/modules/alsa/alsa-ucm.c
@@ -40,6 +40,7 @@
 
 #include <pulsecore/log.h>
 #include <pulsecore/macro.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/atomic.h>
 #include <pulsecore/core-error.h>
diff --git a/src/modules/alsa/module-alsa-card.c b/src/modules/alsa/module-alsa-card.c
index 4f1236e..b6ac865 100644
--- a/src/modules/alsa/module-alsa-card.c
+++ b/src/modules/alsa/module-alsa-card.c
@@ -23,6 +23,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/bluetooth/backend-native.c b/src/modules/bluetooth/backend-native.c
index cf88126..f78cebc 100644
--- a/src/modules/bluetooth/backend-native.c
+++ b/src/modules/bluetooth/backend-native.c
@@ -22,6 +22,7 @@
 #endif
 
 #include <pulsecore/shared.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-shared.h>
diff --git a/src/modules/bluetooth/bluez4-util.c b/src/modules/bluetooth/bluez4-util.c
index 542ce35..c582b95 100644
--- a/src/modules/bluetooth/bluez4-util.c
+++ b/src/modules/bluetooth/bluez4-util.c
@@ -23,6 +23,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/shared.h>
 #include <pulsecore/dbus-shared.h>
diff --git a/src/modules/bluetooth/bluez5-util.c b/src/modules/bluetooth/bluez5-util.c
index 7d63f35..dac176b 100644
--- a/src/modules/bluetooth/bluez5-util.c
+++ b/src/modules/bluetooth/bluez5-util.c
@@ -26,6 +26,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-shared.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/bluetooth/module-bluetooth-policy.c b/src/modules/bluetooth/module-bluetooth-policy.c
index df702cc..aa9fa4a 100644
--- a/src/modules/bluetooth/module-bluetooth-policy.c
+++ b/src/modules/bluetooth/module-bluetooth-policy.c
@@ -26,6 +26,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/source-output.h>
 #include <pulsecore/source.h>
diff --git a/src/modules/bluetooth/module-bluez4-device.c b/src/modules/bluetooth/module-bluez4-device.c
index ac4ed63..3f2e70f 100644
--- a/src/modules/bluetooth/module-bluez4-device.c
+++ b/src/modules/bluetooth/module-bluez4-device.c
@@ -36,6 +36,7 @@
 #include <pulsecore/i18n.h>
 #include <pulsecore/module.h>
 #include <pulsecore/modargs.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/core-error.h>
diff --git a/src/modules/bluetooth/module-bluez5-device.c b/src/modules/bluetooth/module-bluez5-device.c
index 065fcaa..0dbe6f3 100644
--- a/src/modules/bluetooth/module-bluez5-device.c
+++ b/src/modules/bluetooth/module-bluez5-device.c
@@ -30,6 +30,7 @@
 #include <pulse/rtclock.h>
 #include <pulse/timeval.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/bluetooth/module-bluez5-discover.c b/src/modules/bluetooth/module-bluez5-discover.c
index 080e5d0..a5a0b45 100644
--- a/src/modules/bluetooth/module-bluez5-discover.c
+++ b/src/modules/bluetooth/module-bluez5-discover.c
@@ -22,6 +22,7 @@
 #endif
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/macro.h>
 #include <pulsecore/module.h>
diff --git a/src/modules/dbus/iface-card.c b/src/modules/dbus/iface-card.c
index 1b705ba..ffcbc7a 100644
--- a/src/modules/dbus/iface-card.c
+++ b/src/modules/dbus/iface-card.c
@@ -23,6 +23,7 @@
 
 #include <dbus/dbus.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/protocol-dbus.h>
diff --git a/src/modules/dbus/iface-client.c b/src/modules/dbus/iface-client.c
index 455ea45..3f3003f 100644
--- a/src/modules/dbus/iface-client.c
+++ b/src/modules/dbus/iface-client.c
@@ -24,6 +24,7 @@
 
 #include <dbus/dbus.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/protocol-dbus.h>
diff --git a/src/modules/dbus/iface-core.c b/src/modules/dbus/iface-core.c
index 508913d..bfabd61 100644
--- a/src/modules/dbus/iface-core.c
+++ b/src/modules/dbus/iface-core.c
@@ -28,6 +28,7 @@
 #include <pulse/utf8.h>
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/macro.h>
diff --git a/src/modules/dbus/iface-device-port.c b/src/modules/dbus/iface-device-port.c
index 4892443..96a2abd 100644
--- a/src/modules/dbus/iface-device-port.c
+++ b/src/modules/dbus/iface-device-port.c
@@ -23,6 +23,7 @@
 
 #include <dbus/dbus.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 
diff --git a/src/modules/dbus/iface-device.c b/src/modules/dbus/iface-device.c
index 2c370a8..abaa42b 100644
--- a/src/modules/dbus/iface-device.c
+++ b/src/modules/dbus/iface-device.c
@@ -21,6 +21,7 @@
 #include <config.h>
 #endif
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/protocol-dbus.h>
diff --git a/src/modules/dbus/iface-memstats.c b/src/modules/dbus/iface-memstats.c
index cd1d7ea..6ed4b07 100644
--- a/src/modules/dbus/iface-memstats.c
+++ b/src/modules/dbus/iface-memstats.c
@@ -23,6 +23,7 @@
 
 #include <dbus/dbus.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
diff --git a/src/modules/dbus/iface-module.c b/src/modules/dbus/iface-module.c
index 222cd73..a6fdf50 100644
--- a/src/modules/dbus/iface-module.c
+++ b/src/modules/dbus/iface-module.c
@@ -21,6 +21,7 @@
 #include <config.h>
 #endif
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/dbus/iface-sample.c b/src/modules/dbus/iface-sample.c
index 61f2ba0..accdf46 100644
--- a/src/modules/dbus/iface-sample.c
+++ b/src/modules/dbus/iface-sample.c
@@ -21,6 +21,7 @@
 #include <config.h>
 #endif
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/namereg.h>
diff --git a/src/modules/dbus/iface-stream.c b/src/modules/dbus/iface-stream.c
index ade62ca..76bbeb0 100644
--- a/src/modules/dbus/iface-stream.c
+++ b/src/modules/dbus/iface-stream.c
@@ -22,6 +22,7 @@
 #include <config.h>
 #endif
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/protocol-dbus.h>
diff --git a/src/modules/dbus/module-dbus-protocol.c b/src/modules/dbus/module-dbus-protocol.c
index 8079d6b..abd340a 100644
--- a/src/modules/dbus/module-dbus-protocol.c
+++ b/src/modules/dbus/module-dbus-protocol.c
@@ -30,6 +30,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/client.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/dbus-util.h>
 #include <pulsecore/idxset.h>
diff --git a/src/modules/echo-cancel/adrian.c b/src/modules/echo-cancel/adrian.c
index 3c47fae..870a1fe 100644
--- a/src/modules/echo-cancel/adrian.c
+++ b/src/modules/echo-cancel/adrian.c
@@ -29,6 +29,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/modargs.h>
 
 #include "echo-cancel.h"
diff --git a/src/modules/echo-cancel/module-echo-cancel.c b/src/modules/echo-cancel/module-echo-cancel.c
index dfd05b6..f3f24ed 100644
--- a/src/modules/echo-cancel/module-echo-cancel.c
+++ b/src/modules/echo-cancel/module-echo-cancel.c
@@ -43,6 +43,7 @@
 #include <pulsecore/namereg.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/gconf/module-gconf.c b/src/modules/gconf/module-gconf.c
index 1e1b855..5baa1ae 100644
--- a/src/modules/gconf/module-gconf.c
+++ b/src/modules/gconf/module-gconf.c
@@ -32,6 +32,7 @@
 #include <pulse/xmalloc.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/log.h>
 #include <pulse/mainloop-api.h>
diff --git a/src/modules/jack/module-jack-sink.c b/src/modules/jack/module-jack-sink.c
index 4d31493..22dbbad 100644
--- a/src/modules/jack/module-jack-sink.c
+++ b/src/modules/jack/module-jack-sink.c
@@ -33,6 +33,7 @@
 
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/jack/module-jack-source.c b/src/modules/jack/module-jack-source.c
index e45f304..ce5a096 100644
--- a/src/modules/jack/module-jack-source.c
+++ b/src/modules/jack/module-jack-source.c
@@ -33,6 +33,7 @@
 
 #include <pulsecore/source.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-access.c b/src/modules/module-access.c
index c00612a..39e2f0e 100644
--- a/src/modules/module-access.c
+++ b/src/modules/module-access.c
@@ -34,6 +34,7 @@
 #include <pulse/timeval.h>
 
 #include <pulsecore/core-error.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/module-allow-passthrough.c b/src/modules/module-allow-passthrough.c
index 4b801e4..45e7a41 100644
--- a/src/modules/module-allow-passthrough.c
+++ b/src/modules/module-allow-passthrough.c
@@ -26,6 +26,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/source-output.h>
diff --git a/src/modules/module-always-sink.c b/src/modules/module-always-sink.c
index 12f63f7..a7f3568 100644
--- a/src/modules/module-always-sink.c
+++ b/src/modules/module-always-sink.c
@@ -24,6 +24,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/sink.h>
diff --git a/src/modules/module-augment-properties.c b/src/modules/module-augment-properties.c
index 541f0e7..4608d45 100644
--- a/src/modules/module-augment-properties.c
+++ b/src/modules/module-augment-properties.c
@@ -28,6 +28,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-card-restore.c b/src/modules/module-card-restore.c
index 3c0307b..ff6066f 100644
--- a/src/modules/module-card-restore.c
+++ b/src/modules/module-card-restore.c
@@ -33,6 +33,7 @@
 #include <pulse/timeval.h>
 #include <pulse/rtclock.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/module-cli.c b/src/modules/module-cli.c
index a69ad4a..b23519d 100644
--- a/src/modules/module-cli.c
+++ b/src/modules/module-cli.c
@@ -33,6 +33,7 @@
 #include <pulsecore/log.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/macro.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 
 #include "module-cli-symdef.h"
diff --git a/src/modules/module-combine-sink.c b/src/modules/module-combine-sink.c
index 250240a..4fadbb9 100644
--- a/src/modules/module-combine-sink.c
+++ b/src/modules/module-combine-sink.c
@@ -35,6 +35,7 @@
 #include <pulsecore/sink-input.h>
 #include <pulsecore/memblockq.h>
 #include <pulsecore/log.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/module-default-device-restore.c b/src/modules/module-default-device-restore.c
index d76e28e..4b169bb 100644
--- a/src/modules/module-default-device-restore.c
+++ b/src/modules/module-default-device-restore.c
@@ -28,6 +28,7 @@
 #include <pulse/timeval.h>
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/module.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-device-manager.c b/src/modules/module-device-manager.c
index 2a0a67f..17406c3 100644
--- a/src/modules/module-device-manager.c
+++ b/src/modules/module-device-manager.c
@@ -34,6 +34,7 @@
 #include <pulse/timeval.h>
 #include <pulse/rtclock.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/module-device-restore.c b/src/modules/module-device-restore.c
index 8d7b34b..f150dbd 100644
--- a/src/modules/module-device-restore.c
+++ b/src/modules/module-device-restore.c
@@ -37,6 +37,7 @@
 #include <pulse/format.h>
 #include <pulse/internal.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/module-equalizer-sink.c b/src/modules/module-equalizer-sink.c
index 9c25f3f..8c50468 100644
--- a/src/modules/module-equalizer-sink.c
+++ b/src/modules/module-equalizer-sink.c
@@ -46,6 +46,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/timeval.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/aupdate.h>
diff --git a/src/modules/module-esound-sink.c b/src/modules/module-esound-sink.c
index 2ce0c85..f304c97 100644
--- a/src/modules/module-esound-sink.c
+++ b/src/modules/module-esound-sink.c
@@ -48,6 +48,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/socket.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/iochannel.h>
 #include <pulsecore/sink.h>
diff --git a/src/modules/module-filter-apply.c b/src/modules/module-filter-apply.c
index 364d68b..ceeb541 100644
--- a/src/modules/module-filter-apply.c
+++ b/src/modules/module-filter-apply.c
@@ -26,6 +26,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/macro.h>
diff --git a/src/modules/module-filter-heuristics.c b/src/modules/module-filter-heuristics.c
index 423e873..15baa1c 100644
--- a/src/modules/module-filter-heuristics.c
+++ b/src/modules/module-filter-heuristics.c
@@ -26,6 +26,7 @@
 #include <pulsecore/macro.h>
 #include <pulsecore/hook-list.h>
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/module-intended-roles.c b/src/modules/module-intended-roles.c
index 60ec7e9..e083c11 100644
--- a/src/modules/module-intended-roles.c
+++ b/src/modules/module-intended-roles.c
@@ -24,6 +24,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-ladspa-sink.c b/src/modules/module-ladspa-sink.c
index c11fa5e..3602830 100644
--- a/src/modules/module-ladspa-sink.c
+++ b/src/modules/module-ladspa-sink.c
@@ -32,6 +32,7 @@
 #include <pulsecore/namereg.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-lirc.c b/src/modules/module-lirc.c
index d63fdbd..7c325da 100644
--- a/src/modules/module-lirc.c
+++ b/src/modules/module-lirc.c
@@ -30,6 +30,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/module.h>
 #include <pulsecore/log.h>
 #include <pulsecore/namereg.h>
diff --git a/src/modules/module-loopback.c b/src/modules/module-loopback.c
index 9410c98..75da911 100644
--- a/src/modules/module-loopback.c
+++ b/src/modules/module-loopback.c
@@ -31,6 +31,7 @@
 #include <pulsecore/modargs.h>
 #include <pulsecore/namereg.h>
 #include <pulsecore/log.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 
 #include <pulse/rtclock.h>
diff --git a/src/modules/module-match.c b/src/modules/module-match.c
index 559687c..f371139 100644
--- a/src/modules/module-match.c
+++ b/src/modules/module-match.c
@@ -36,6 +36,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/module-mmkbd-evdev.c b/src/modules/module-mmkbd-evdev.c
index 6c9cf34..818a502 100644
--- a/src/modules/module-mmkbd-evdev.c
+++ b/src/modules/module-mmkbd-evdev.c
@@ -32,6 +32,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-native-protocol-fd.c b/src/modules/module-native-protocol-fd.c
index 5fd7f6a..c951bc6 100644
--- a/src/modules/module-native-protocol-fd.c
+++ b/src/modules/module-native-protocol-fd.c
@@ -24,6 +24,7 @@
 #include <stdio.h>
 #include <unistd.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/module.h>
 #include <pulsecore/macro.h>
 #include <pulsecore/iochannel.h>
diff --git a/src/modules/module-null-sink.c b/src/modules/module-null-sink.c
index b8157e8..be3c98f 100644
--- a/src/modules/module-null-sink.c
+++ b/src/modules/module-null-sink.c
@@ -34,6 +34,7 @@
 #include <pulsecore/macro.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-null-source.c b/src/modules/module-null-source.c
index a75a04f..a791883 100644
--- a/src/modules/module-null-source.c
+++ b/src/modules/module-null-source.c
@@ -31,6 +31,7 @@
 #include <pulse/timeval.h>
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/log.h>
 #include <pulsecore/macro.h>
diff --git a/src/modules/module-pipe-sink.c b/src/modules/module-pipe-sink.c
index da65021..0b32f0d 100644
--- a/src/modules/module-pipe-sink.c
+++ b/src/modules/module-pipe-sink.c
@@ -35,6 +35,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
diff --git a/src/modules/module-pipe-source.c b/src/modules/module-pipe-source.c
index f39fc55..4b651cc 100644
--- a/src/modules/module-pipe-source.c
+++ b/src/modules/module-pipe-source.c
@@ -35,6 +35,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/source.h>
 #include <pulsecore/module.h>
diff --git a/src/modules/module-position-event-sounds.c b/src/modules/module-position-event-sounds.c
index fb9f5b3..ef03786 100644
--- a/src/modules/module-position-event-sounds.c
+++ b/src/modules/module-position-event-sounds.c
@@ -32,6 +32,7 @@
 #include <pulse/channelmap.h>
 
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-protocol-stub.c b/src/modules/module-protocol-stub.c
index 5a183c8..8fd4392 100644
--- a/src/modules/module-protocol-stub.c
+++ b/src/modules/module-protocol-stub.c
@@ -32,6 +32,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/socket.h>
diff --git a/src/modules/module-rescue-streams.c b/src/modules/module-rescue-streams.c
index 60ac1c4..e844e16 100644
--- a/src/modules/module-rescue-streams.c
+++ b/src/modules/module-rescue-streams.c
@@ -24,6 +24,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/source-output.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/module-role-cork.c b/src/modules/module-role-cork.c
index d71cc38..d11be8c 100644
--- a/src/modules/module-role-cork.c
+++ b/src/modules/module-role-cork.c
@@ -23,6 +23,7 @@
 
 #include <pulsecore/macro.h>
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <stream-interaction.h>
 
 #include "module-role-cork-symdef.h"
diff --git a/src/modules/module-role-ducking.c b/src/modules/module-role-ducking.c
index add2d36..af3e98c 100644
--- a/src/modules/module-role-ducking.c
+++ b/src/modules/module-role-ducking.c
@@ -23,6 +23,7 @@
 
 #include <pulsecore/macro.h>
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <stream-interaction.h>
 
 #include "module-role-ducking-symdef.h"
diff --git a/src/modules/module-rygel-media-server.c b/src/modules/module-rygel-media-server.c
index e2c2e6f..0737fa9 100644
--- a/src/modules/module-rygel-media-server.c
+++ b/src/modules/module-rygel-media-server.c
@@ -33,6 +33,7 @@
 #include <pulsecore/i18n.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/source.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/log.h>
 #include <pulsecore/macro.h>
diff --git a/src/modules/module-sine-source.c b/src/modules/module-sine-source.c
index cdeb2c0..a3be322 100644
--- a/src/modules/module-sine-source.c
+++ b/src/modules/module-sine-source.c
@@ -32,6 +32,7 @@
 
 #include <pulsecore/source.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-sine.c b/src/modules/module-sine.c
index d56fae5..a95cf79 100644
--- a/src/modules/module-sine.c
+++ b/src/modules/module-sine.c
@@ -25,6 +25,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/module.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/module-stream-restore.c b/src/modules/module-stream-restore.c
index d1c2597..75ca729 100644
--- a/src/modules/module-stream-restore.c
+++ b/src/modules/module-stream-restore.c
@@ -35,6 +35,7 @@
 #include <pulse/timeval.h>
 #include <pulse/rtclock.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/module-suspend-on-idle.c b/src/modules/module-suspend-on-idle.c
index a284f85..030de4d 100644
--- a/src/modules/module-suspend-on-idle.c
+++ b/src/modules/module-suspend-on-idle.c
@@ -26,6 +26,7 @@
 #include <pulse/rtclock.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/source-output.h>
diff --git a/src/modules/module-switch-on-connect.c b/src/modules/module-switch-on-connect.c
index c844add..f3171e9 100644
--- a/src/modules/module-switch-on-connect.c
+++ b/src/modules/module-switch-on-connect.c
@@ -25,6 +25,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/source-output.h>
 #include <pulsecore/source.h>
diff --git a/src/modules/module-switch-on-port-available.c b/src/modules/module-switch-on-port-available.c
index b9a0f3b..ee787cc 100644
--- a/src/modules/module-switch-on-port-available.c
+++ b/src/modules/module-switch-on-port-available.c
@@ -23,6 +23,7 @@
 #endif
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/device-port.h>
 #include <pulsecore/hashmap.h>
diff --git a/src/modules/module-systemd-login.c b/src/modules/module-systemd-login.c
index d15bee5..62eeb4c 100644
--- a/src/modules/module-systemd-login.c
+++ b/src/modules/module-systemd-login.c
@@ -32,6 +32,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/module.h>
 #include <pulsecore/log.h>
 #include <pulsecore/hashmap.h>
diff --git a/src/modules/module-tunnel-sink-new.c b/src/modules/module-tunnel-sink-new.c
index 92f99df..82eb7c6 100644
--- a/src/modules/module-tunnel-sink-new.c
+++ b/src/modules/module-tunnel-sink-new.c
@@ -30,6 +30,7 @@
 #include <pulse/error.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/sink.h>
diff --git a/src/modules/module-tunnel-source-new.c b/src/modules/module-tunnel-source-new.c
index e159c33..ff40140 100644
--- a/src/modules/module-tunnel-source-new.c
+++ b/src/modules/module-tunnel-source-new.c
@@ -30,6 +30,7 @@
 #include <pulse/error.h>
 
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/i18n.h>
 #include <pulsecore/source.h>
diff --git a/src/modules/module-tunnel.c b/src/modules/module-tunnel.c
index d7114ee..9e5f316 100644
--- a/src/modules/module-tunnel.c
+++ b/src/modules/module-tunnel.c
@@ -40,6 +40,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-udev-detect.c b/src/modules/module-udev-detect.c
index bb41a96..d36b136 100644
--- a/src/modules/module-udev-detect.c
+++ b/src/modules/module-udev-detect.c
@@ -30,6 +30,7 @@
 #include <pulse/timeval.h>
 
 #include <pulsecore/modargs.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/namereg.h>
diff --git a/src/modules/module-virtual-sink.c b/src/modules/module-virtual-sink.c
index 02cc1ac..286310e 100644
--- a/src/modules/module-virtual-sink.c
+++ b/src/modules/module-virtual-sink.c
@@ -29,6 +29,7 @@
 #include <pulsecore/namereg.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-virtual-source.c b/src/modules/module-virtual-source.c
index 36edf78..b3403a6 100644
--- a/src/modules/module-virtual-source.c
+++ b/src/modules/module-virtual-source.c
@@ -31,6 +31,7 @@
 #include <pulsecore/namereg.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-virtual-surround-sink.c b/src/modules/module-virtual-surround-sink.c
index 6c7120a..4d0cee5 100644
--- a/src/modules/module-virtual-surround-sink.c
+++ b/src/modules/module-virtual-surround-sink.c
@@ -30,6 +30,7 @@
 #include <pulsecore/namereg.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
diff --git a/src/modules/module-zeroconf-discover.c b/src/modules/module-zeroconf-discover.c
index 96476b7..e628c6b 100644
--- a/src/modules/module-zeroconf-discover.c
+++ b/src/modules/module-zeroconf-discover.c
@@ -35,6 +35,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/log.h>
 #include <pulsecore/hashmap.h>
diff --git a/src/modules/module-zeroconf-publish.c b/src/modules/module-zeroconf-publish.c
index 482881c..74018eb 100644
--- a/src/modules/module-zeroconf-publish.c
+++ b/src/modules/module-zeroconf-publish.c
@@ -39,6 +39,7 @@
 #include <pulsecore/sink.h>
 #include <pulsecore/source.h>
 #include <pulsecore/native-common.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/log.h>
 #include <pulsecore/dynarray.h>
diff --git a/src/modules/oss/module-oss.c b/src/modules/oss/module-oss.c
index 8a5a692..a53b725 100644
--- a/src/modules/oss/module-oss.c
+++ b/src/modules/oss/module-oss.c
@@ -51,6 +51,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/util.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/thread.h>
 #include <pulsecore/sink.h>
diff --git a/src/modules/raop/module-raop-discover.c b/src/modules/raop/module-raop-discover.c
index f083044..f4ac133 100644
--- a/src/modules/raop/module-raop-discover.c
+++ b/src/modules/raop/module-raop-discover.c
@@ -36,6 +36,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/log.h>
 #include <pulsecore/hashmap.h>
diff --git a/src/modules/raop/module-raop-sink.c b/src/modules/raop/module-raop-sink.c
index 7a97e83..2827b04 100644
--- a/src/modules/raop/module-raop-sink.c
+++ b/src/modules/raop/module-raop-sink.c
@@ -41,6 +41,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/core-error.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/sink.h>
 #include <pulsecore/module.h>
 #include <pulsecore/core-util.h>
diff --git a/src/modules/raop/raop_client.c b/src/modules/raop/raop_client.c
index 864c558..bd4f90b 100644
--- a/src/modules/raop/raop_client.c
+++ b/src/modules/raop/raop_client.c
@@ -40,6 +40,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/iochannel.h>
diff --git a/src/modules/rtp/module-rtp-recv.c b/src/modules/rtp/module-rtp-recv.c
index 5977500..e77f3d0 100644
--- a/src/modules/rtp/module-rtp-recv.c
+++ b/src/modules/rtp/module-rtp-recv.c
@@ -34,6 +34,7 @@
 #include <pulse/timeval.h>
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/llist.h>
diff --git a/src/modules/rtp/module-rtp-send.c b/src/modules/rtp/module-rtp-send.c
index 15a7436..ef5c526 100644
--- a/src/modules/rtp/module-rtp-send.c
+++ b/src/modules/rtp/module-rtp-send.c
@@ -32,6 +32,7 @@
 #include <pulse/util.h>
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/module.h>
 #include <pulsecore/source.h>
diff --git a/src/modules/stream-interaction.c b/src/modules/stream-interaction.c
index 4184786..728398c 100644
--- a/src/modules/stream-interaction.c
+++ b/src/modules/stream-interaction.c
@@ -28,6 +28,7 @@
 #include <pulsecore/hashmap.h>
 #include <pulsecore/hook-list.h>
 #include <pulsecore/core.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/modargs.h>
diff --git a/src/modules/x11/module-x11-cork-request.c b/src/modules/x11/module-x11-cork-request.c
index 5c76711..a86a82e 100644
--- a/src/modules/x11/module-x11-cork-request.c
+++ b/src/modules/x11/module-x11-cork-request.c
@@ -36,6 +36,7 @@
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
 #include <pulsecore/x11wrap.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 
 #include "module-x11-cork-request-symdef.h"
diff --git a/src/modules/x11/module-x11-xsmp.c b/src/modules/x11/module-x11-xsmp.c
index 7c6fb23..5f09d52 100644
--- a/src/modules/x11/module-x11-xsmp.c
+++ b/src/modules/x11/module-x11-xsmp.c
@@ -31,6 +31,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/util.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/modargs.h>
 #include <pulsecore/log.h>
 #include <pulsecore/x11wrap.h>
diff --git a/src/pulsecore/card.c b/src/pulsecore/card.c
index bc5b75b..39f4261 100644
--- a/src/pulsecore/card.c
+++ b/src/pulsecore/card.c
@@ -30,6 +30,7 @@
 
 #include <pulsecore/log.h>
 #include <pulsecore/macro.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/namereg.h>
 #include <pulsecore/device-port.h>
diff --git a/src/pulsecore/cli-command.c b/src/pulsecore/cli-command.c
index 9a73605..05c2b8f 100644
--- a/src/pulsecore/cli-command.c
+++ b/src/pulsecore/cli-command.c
@@ -47,6 +47,7 @@
 #include <pulsecore/strbuf.h>
 #include <pulsecore/namereg.h>
 #include <pulsecore/cli-text.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/sound-file.h>
 #include <pulsecore/play-memchunk.h>
diff --git a/src/pulsecore/cli-text.c b/src/pulsecore/cli-text.c
index af79a1e..9e32524 100644
--- a/src/pulsecore/cli-text.c
+++ b/src/pulsecore/cli-text.c
@@ -32,6 +32,7 @@
 #include <pulsecore/sink-input.h>
 #include <pulsecore/source-output.h>
 #include <pulsecore/strbuf.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/macro.h>
 #include <pulsecore/core-util.h>
diff --git a/src/pulsecore/client.c b/src/pulsecore/client.c
index 2e6af47..24faa1f 100644
--- a/src/pulsecore/client.c
+++ b/src/pulsecore/client.c
@@ -29,6 +29,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/util.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-subscribe.h>
 #include <pulsecore/log.h>
 #include <pulsecore/macro.h>
diff --git a/src/pulsecore/core-scache.c b/src/pulsecore/core-scache.c
index 026eeca..8d54720 100644
--- a/src/pulsecore/core-scache.c
+++ b/src/pulsecore/core-scache.c
@@ -49,6 +49,7 @@
 
 #include <pulsecore/sink-input.h>
 #include <pulsecore/play-memchunk.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-subscribe.h>
 #include <pulsecore/namereg.h>
 #include <pulsecore/sound-file.h>
diff --git a/src/pulsecore/core-struct.h b/src/pulsecore/core-struct.h
new file mode 100644
index 0000000..e6aa374
--- /dev/null
+++ b/src/pulsecore/core-struct.h
@@ -0,0 +1,103 @@
+#ifndef foocorestructhfoo
+#define foocorestructhfoo
+
+/***
+  This file is part of PulseAudio.
+
+  Copyright 2004-2006 Lennart Poettering
+
+  PulseAudio is free software; you can redistribute it and/or modify
+  it under the terms of the GNU Lesser General Public License as published
+  by the Free Software Foundation; either version 2.1 of the License,
+  or (at your option) any later version.
+
+  PulseAudio is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include "core.h"
+
+/* The core structure of PulseAudio. Every PulseAudio daemon contains
+ * exactly one of these. It is used for storing kind of global
+ * variables for the daemon. */
+
+struct pa_core {
+    pa_msgobject parent;
+
+    pa_core_state_t state;
+
+    /* A random value which may be used to identify this instance of
+     * PulseAudio. Not cryptographically secure in any way. */
+    uint32_t cookie;
+
+    pa_mainloop_api *mainloop;
+
+    /* idxset of all kinds of entities */
+    pa_idxset *clients, *cards, *sinks, *sources, *sink_inputs, *source_outputs, *modules, *scache;
+
+    /* Some hashmaps for all sorts of entities */
+    pa_hashmap *namereg, *shared;
+
+    /* The default sink/source */
+    pa_source *default_source;
+    pa_sink *default_sink;
+
+    pa_channel_map default_channel_map;
+    pa_sample_spec default_sample_spec;
+    uint32_t alternate_sample_rate;
+    unsigned default_n_fragments, default_fragment_size_msec;
+    unsigned deferred_volume_safety_margin_usec;
+    int deferred_volume_extra_delay_usec;
+    unsigned lfe_crossover_freq;
+
+    pa_defer_event *module_defer_unload_event;
+    pa_hashmap *modules_pending_unload; /* pa_module -> pa_module (hashmap-as-a-set) */
+
+    pa_defer_event *subscription_defer_event;
+    PA_LLIST_HEAD(pa_subscription, subscriptions);
+    PA_LLIST_HEAD(pa_subscription_event, subscription_event_queue);
+    pa_subscription_event *subscription_event_last;
+
+    /* The mempool is used for data we write to, it's readonly for the client. */
+    pa_mempool *mempool;
+
+    /* Shared memory size, as specified either by daemon configuration
+     * or PA daemon defaults (~ 64 MiB). */
+    size_t shm_size;
+
+    pa_silence_cache silence_cache;
+
+    pa_time_event *exit_event;
+    pa_time_event *scache_auto_unload_event;
+
+    int exit_idle_time, scache_idle_time;
+
+    bool flat_volumes:1;
+    bool disallow_module_loading:1;
+    bool disallow_exit:1;
+    bool running_as_daemon:1;
+    bool realtime_scheduling:1;
+    bool avoid_resampling:1;
+    bool disable_remixing:1;
+    bool remixing_use_all_sink_channels:1;
+    bool disable_lfe_remixing:1;
+    bool deferred_volume:1;
+
+    pa_resample_method_t resample_method;
+    int realtime_priority;
+
+    pa_server_type_t server_type;
+    pa_cpu_info cpu_info;
+
+    /* hooks */
+    pa_hook hooks[PA_CORE_HOOK_MAX];
+    /* access hooks */
+    pa_hook access[PA_ACCESS_HOOK_MAX];
+};
+
+#endif /* foocorestructhfoo */
diff --git a/src/pulsecore/core-subscribe.c b/src/pulsecore/core-subscribe.c
index 5a88b7e..88d900b 100644
--- a/src/pulsecore/core-subscribe.c
+++ b/src/pulsecore/core-subscribe.c
@@ -28,6 +28,7 @@
 #include <pulsecore/log.h>
 #include <pulsecore/macro.h>
 
+#include <pulsecore/core-struct.h>
 #include "core-subscribe.h"
 
 /* The subscription subsystem may be used to be notified whenever an
diff --git a/src/pulsecore/core.c b/src/pulsecore/core.c
index e5d6879..b8dbde9 100644
--- a/src/pulsecore/core.c
+++ b/src/pulsecore/core.c
@@ -31,6 +31,7 @@
 #include <pulse/xmalloc.h>
 
 #include <pulsecore/module.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/core-scache.h>
diff --git a/src/pulsecore/core.h b/src/pulsecore/core.h
index 152a53f..955db89 100644
--- a/src/pulsecore/core.h
+++ b/src/pulsecore/core.h
@@ -141,82 +141,6 @@ typedef enum pa_core_hook {
     PA_CORE_HOOK_MAX
 } pa_core_hook_t;
 
-/* The core structure of PulseAudio. Every PulseAudio daemon contains
- * exactly one of these. It is used for storing kind of global
- * variables for the daemon. */
-
-struct pa_core {
-    pa_msgobject parent;
-
-    pa_core_state_t state;
-
-    /* A random value which may be used to identify this instance of
-     * PulseAudio. Not cryptographically secure in any way. */
-    uint32_t cookie;
-
-    pa_mainloop_api *mainloop;
-
-    /* idxset of all kinds of entities */
-    pa_idxset *clients, *cards, *sinks, *sources, *sink_inputs, *source_outputs, *modules, *scache;
-
-    /* Some hashmaps for all sorts of entities */
-    pa_hashmap *namereg, *shared;
-
-    /* The default sink/source */
-    pa_source *default_source;
-    pa_sink *default_sink;
-
-    pa_channel_map default_channel_map;
-    pa_sample_spec default_sample_spec;
-    uint32_t alternate_sample_rate;
-    unsigned default_n_fragments, default_fragment_size_msec;
-    unsigned deferred_volume_safety_margin_usec;
-    int deferred_volume_extra_delay_usec;
-    unsigned lfe_crossover_freq;
-
-    pa_defer_event *module_defer_unload_event;
-    pa_hashmap *modules_pending_unload; /* pa_module -> pa_module (hashmap-as-a-set) */
-
-    pa_defer_event *subscription_defer_event;
-    PA_LLIST_HEAD(pa_subscription, subscriptions);
-    PA_LLIST_HEAD(pa_subscription_event, subscription_event_queue);
-    pa_subscription_event *subscription_event_last;
-
-    /* The mempool is used for data we write to, it's readonly for the client. */
-    pa_mempool *mempool;
-
-    /* Shared memory size, as specified either by daemon configuration
-     * or PA daemon defaults (~ 64 MiB). */
-    size_t shm_size;
-
-    pa_silence_cache silence_cache;
-
-    pa_time_event *exit_event;
-    pa_time_event *scache_auto_unload_event;
-
-    int exit_idle_time, scache_idle_time;
-
-    bool flat_volumes:1;
-    bool disallow_module_loading:1;
-    bool disallow_exit:1;
-    bool running_as_daemon:1;
-    bool realtime_scheduling:1;
-    bool disable_remixing:1;
-    bool disable_lfe_remixing:1;
-    bool deferred_volume:1;
-
-    pa_resample_method_t resample_method;
-    int realtime_priority;
-
-    pa_server_type_t server_type;
-    pa_cpu_info cpu_info;
-
-    /* hooks */
-    pa_hook hooks[PA_CORE_HOOK_MAX];
-    /* access hooks */
-    pa_hook access[PA_ACCESS_HOOK_MAX];
-};
-
 PA_DECLARE_PUBLIC_CLASS(pa_core);
 #define PA_CORE(o) pa_core_cast(o)
 
diff --git a/src/pulsecore/dbus-shared.c b/src/pulsecore/dbus-shared.c
index 935b068..3309f87 100644
--- a/src/pulsecore/dbus-shared.c
+++ b/src/pulsecore/dbus-shared.c
@@ -24,6 +24,7 @@
 
 #include <pulse/xmalloc.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/shared.h>
 
 #include "dbus-shared.h"
diff --git a/src/pulsecore/device-port.c b/src/pulsecore/device-port.c
index 7c9ddf3..bc4d704 100644
--- a/src/pulsecore/device-port.c
+++ b/src/pulsecore/device-port.c
@@ -21,6 +21,7 @@
 
 #include "device-port.h"
 #include <pulsecore/card.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 
 PA_DEFINE_PUBLIC_CLASS(pa_device_port, pa_object);
diff --git a/src/pulsecore/module.c b/src/pulsecore/module.c
index ac15815..da2abe0 100644
--- a/src/pulsecore/module.c
+++ b/src/pulsecore/module.c
@@ -31,6 +31,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/proplist.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-subscribe.h>
 #include <pulsecore/log.h>
 #include <pulsecore/core-util.h>
diff --git a/src/pulsecore/namereg.c b/src/pulsecore/namereg.c
index 47bfc08..dfdcc43 100644
--- a/src/pulsecore/namereg.c
+++ b/src/pulsecore/namereg.c
@@ -30,6 +30,7 @@
 
 #include <pulsecore/source.h>
 #include <pulsecore/sink.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-subscribe.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/macro.h>
diff --git a/src/pulsecore/play-memchunk.c b/src/pulsecore/play-memchunk.c
index cc0bc16..3c7e5a6 100644
--- a/src/pulsecore/play-memchunk.c
+++ b/src/pulsecore/play-memchunk.c
@@ -24,6 +24,7 @@
 #include <stdlib.h>
 #include <stdio.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/play-memblockq.h>
 
diff --git a/src/pulsecore/protocol-esound.c b/src/pulsecore/protocol-esound.c
index 4f83ac4..ead1d81 100644
--- a/src/pulsecore/protocol-esound.c
+++ b/src/pulsecore/protocol-esound.c
@@ -41,6 +41,7 @@
 #include <pulsecore/sink.h>
 #include <pulsecore/source-output.h>
 #include <pulsecore/source.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/sample-util.h>
 #include <pulsecore/namereg.h>
diff --git a/src/pulsecore/protocol-http.c b/src/pulsecore/protocol-http.c
index 25a2cd0..bbb03c3 100644
--- a/src/pulsecore/protocol-http.c
+++ b/src/pulsecore/protocol-http.c
@@ -30,6 +30,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/timeval.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/ioline.h>
 #include <pulsecore/thread-mq.h>
diff --git a/src/pulsecore/protocol-native.c b/src/pulsecore/protocol-native.c
index 8f07b2d..cbb6ae6 100644
--- a/src/pulsecore/protocol-native.c
+++ b/src/pulsecore/protocol-native.c
@@ -45,6 +45,7 @@
 #include <pulsecore/pdispatch.h>
 #include <pulsecore/pstream-util.h>
 #include <pulsecore/namereg.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/core-subscribe.h>
 #include <pulsecore/log.h>
diff --git a/src/pulsecore/protocol-simple.c b/src/pulsecore/protocol-simple.c
index 923ec87..bd9136b 100644
--- a/src/pulsecore/protocol-simple.c
+++ b/src/pulsecore/protocol-simple.c
@@ -34,6 +34,7 @@
 #include <pulsecore/sample-util.h>
 #include <pulsecore/namereg.h>
 #include <pulsecore/log.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/atomic.h>
 #include <pulsecore/thread-mq.h>
diff --git a/src/pulsecore/shared.c b/src/pulsecore/shared.c
index 1b5eea9..2e63d0e 100644
--- a/src/pulsecore/shared.c
+++ b/src/pulsecore/shared.c
@@ -22,6 +22,7 @@
 #endif
 
 #include <pulse/xmalloc.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/macro.h>
 
 #include "shared.h"
diff --git a/src/pulsecore/sink-input.c b/src/pulsecore/sink-input.c
index e9694f2..9a866b4 100644
--- a/src/pulsecore/sink-input.c
+++ b/src/pulsecore/sink-input.c
@@ -30,6 +30,7 @@
 #include <pulse/util.h>
 #include <pulse/internal.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-format.h>
 #include <pulsecore/mix.h>
 #include <pulsecore/stream-util.h>
diff --git a/src/pulsecore/sink.c b/src/pulsecore/sink.c
index aa21822..20249ec 100644
--- a/src/pulsecore/sink.c
+++ b/src/pulsecore/sink.c
@@ -38,6 +38,7 @@
 #include <pulsecore/i18n.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/namereg.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/sample-util.h>
 #include <pulsecore/mix.h>
diff --git a/src/pulsecore/sound-file-stream.c b/src/pulsecore/sound-file-stream.c
index ddda456..3ac935d 100644
--- a/src/pulsecore/sound-file-stream.c
+++ b/src/pulsecore/sound-file-stream.c
@@ -32,6 +32,7 @@
 #include <pulse/xmalloc.h>
 #include <pulse/util.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-error.h>
 #include <pulsecore/sink-input.h>
 #include <pulsecore/log.h>
diff --git a/src/pulsecore/source-output.c b/src/pulsecore/source-output.c
index 6714ea9..a265e6f 100644
--- a/src/pulsecore/source-output.c
+++ b/src/pulsecore/source-output.c
@@ -30,6 +30,7 @@
 #include <pulse/util.h>
 #include <pulse/internal.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-format.h>
 #include <pulsecore/mix.h>
 #include <pulsecore/stream-util.h>
diff --git a/src/pulsecore/source.c b/src/pulsecore/source.c
index 8ce7818..7a33d5b 100644
--- a/src/pulsecore/source.c
+++ b/src/pulsecore/source.c
@@ -33,6 +33,7 @@
 #include <pulse/rtclock.h>
 #include <pulse/internal.h>
 
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/source-output.h>
 #include <pulsecore/namereg.h>
diff --git a/src/pulsecore/x11wrap.c b/src/pulsecore/x11wrap.c
index 0c040cf..76cb83b 100644
--- a/src/pulsecore/x11wrap.c
+++ b/src/pulsecore/x11wrap.c
@@ -28,6 +28,7 @@
 #include <pulsecore/llist.h>
 #include <pulsecore/log.h>
 #include <pulsecore/shared.h>
+#include <pulsecore/core-struct.h>
 #include <pulsecore/core-util.h>
 #include <pulsecore/macro.h>
 
-- 
2.9.3

